src/mem_tracking.rs

```diff
@@ -0,0 +1,86 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{Decode, Error, Input};
+use impl_trait_for_tuples::impl_for_tuples;
+
+/// Marker trait used for identifying types that call the mem tracking hooks exposed by `Input`
+/// while decoding.
+pub trait DecodeWithMemTracking: Decode {}
```

Comment:
I guess it would be great to implement it also on:
* `Compact<T>` where `Compact<T::As>: DecodeWithMemTracking`
* all the explicit compact: `Compact<()>`, `Compact<u8>`...`Compact<u128>`
* stuff in bit_vec module
* stuff in generic_array module

Reply:
Done for all except `generic_array` with no additional changes needed.

With `generic_array` I'm not familiar and I don't know if anyone uses it. In polkadot-sdk it doesn't seem to be used. We could implement it later if someone requested it.

Reply:
Yes we can implement generic_array on demand.